drivers/watchdog/sunxi-twd.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright © 2017 Samuel Holland <samuel@sholland.org>
+ * Copyright © 2018 Drew Walters <drewwalters96@gmail.com>
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <drivers/watchdog/sunxi-twd.h>
+#include <platform/devices.h>
+#include <error.h>
+#include <mmio.h>
+
+/* Default timeout (in seconds) enabled at start. */
+#define TWD_DEFAULT_TIMEOUT 5
+
+static int
+sunxi_twd_probe(struct device *dev __unused)
+{
+	return SUCCESS;
+}
+
+static int
+sunxi_twd_disable(struct device *twd_dev)
+{
+	uint32_t reg_val;
+
+	/* Stop */
+	reg_val = mmio_read32(DEV_R_TWD + R_TWD_CTRL_REG);
+	reg_val |= (1 << 1);
+
+	mmio_write32(DEV_R_TWD + R_TWD_CTRL_REG, reg_val);
+}
+
+static int
+sunxi_twd_enable(struct device *twd_dev)
+{
+	uint32_t reg_val;
+
+	/* Set timeout value to default. */
+	sunxi_twd_set_timeout(twd_dev, TWD_DEFAULT_TIMEOUT);
+
+	/* Reset enable. */
+	reg_val = mmio_read32(DEV_R_TWD + R_TWD_CTRL_REG);
+	reg_val |= (1 << 8);
+
+	/* Enable and perform restart. */
+	reg_val = 1;
+	reg_val |= (0xD14 << 15);
+
+	mmio_write32(DEV_R_TWD + R_TWD_RESTART_REG, reg_val);
+}
+
+static int
+sunxi_twd_set_timeout(struct device *twd_dev, uint8_t timeout)
```

Comment:
This is leftover from the other watchdog, right? Trusted watchdog just uses a linear interval.